azure-iot-device/azure/iot/device/provisioning/internal/request_response_provider.py

```diff
@@ -0,0 +1,80 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+import logging
+import json
+from ..models.registration_result import RegistrationResult, RegistrationState
+
+logger = logging.getLogger(__name__)
+
+
+class RequestResponseMachine(object):
+    def __init__(self, state_based_provider):
+
+        self._state_based_provider = state_based_provider
+        self._state_based_provider.on_state_based_provider_message_received = self._receive_response
+
+        self.on_response_received = None
+
+    def send_request(self, publish_action):
+        """
+        :param publish_action: The publish action
+        """
+        self.publish([publish_action])
+
+    def connect(self, callback=None, post_connect_actions=None):
+        # TODO : MQTT Transport connect should take post connect actions.
+        self._state_based_provider.connect(callback=callback, actions=post_connect_actions)
+
+    def disconnect(self, callback=None, pre_disconnect_actions=None):
+        # TODO : MQTT Transport connect should take pre disconnect actions.
+        self._state_based_provider.disconnect(callback=callback, actions=pre_disconnect_actions)
+
+    def publish(self, publish_actions=None):
+        self._state_based_provider.publish(publish_actions)
+
+    def subscribe(self, subscribe_actions=None):
+        self._state_based_provider.subscribe(subscribe_actions)
+
+    def _receive_response(self, topic, payload):
+        """
+        Handler that processes the response from the service.
+        :param topic: The topic in bytes name on which the message arrived on.
+        :param payload: Payload in bytes of the message received.
+        """
+        # """ Sample topic and payload
+        # $dps/registrations/res/200/?$rid=28c32371-608c-4390-8da7-c712353c1c3b
+        # {"operationId":"4.550cb20c3349a409.390d2957-7b58-4701-b4f9-7fe848348f4a","status":"assigning"}
+        # """
+        logger.info("")
+        topic_str = topic.decode("utf-8")
+        if topic_str.startswith("$dps/registrations/res/"):
+            topic_parts = topic_str.split("$")
+            rid_parts = topic_parts[2].split("=")
+            rid = rid_parts[1]
+
+        payload_str = payload.decode("utf-8")
+        decoded_result = json.loads(payload_str)
+        decoded_state = (
+            None
+            if "registrationState" not in decoded_result
+            else decoded_result["registrationState"]
+        )
+        registration_state = None
+        if decoded_state is not None:
+            registration_state = RegistrationState(
```

Comment:
As you have it here, this layer receives registration state responses.  I think the parsing of the registration state should happen above this layer and this layer shouldn't assume anything about what the payload looks like.

Reply:
are you referring to just registration state or the complete registration result ?

---
In reply to: [277397350](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277397350) [](ancestors = 277397350)

Reply:
both.  I was imagining this layer as knowing nothing about registration.  It only knows about requests and responses, but it doesn't know anything about what is inside those requests or responses.  

---
In reply to: [277443164](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277443164) [](ancestors = 277443164,277397350)

Reply:
it still has to know about the rid from the topic so that it can match it

---
In reply to: [277785920](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277785920) [](ancestors = 277785920,277443164,277397350)